numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2019,6 +2020,80 @@ static PyObject *
     Py_INCREF(ret);
     return ret;
 }
+
+static PyObject *
+@name@_hex(PyObject *self) {
+#if @is_half@
+    npy_float val = npy_half_to_float(PyArrayScalar_VAL(self, @Name@));
+#else
+    npy_@name@ val = PyArrayScalar_VAL(self, @Name@);
+#endif
+    // xref: https://stackoverflow.com/a/15197157/5671364
+    char buf[
+        1 + // sign, '-' or '+'
+#if @is_half@
+        (NPY_SIZEOF_FLOAT * CHAR_BIT + 3) / 4 + // mantissa hex digits max
+#else
+        (NPY_SIZEOF_@NAME@ * CHAR_BIT + 3) / 4 + // mantissa hex digits max
+#endif
+        1 + // decimal point, '.'
+        1 + // mantissa-exponent separator, 'p'
+        1 + // mantissa sign, '-' or '+'
+#if @is_half@
+        (NPY_SIZEOF_FLOAT * CHAR_BIT + 2) / 3 + // exponent decimal digits max
+#else
+        (NPY_SIZEOF_@NAME@ * CHAR_BIT + 2) / 3 + // exponent decimal digits max
+#endif
```

Comment:
There is a staggering amount of `#if @is_half@` :), open to any suggestions to avoid them. We can convert the `PyObject` of `half` itself to `float` and try perhaps?

Reply:
Kinda resolved in 07b4614, not very happy with the solution, hence keeping this thread open to suggestions,